File: src/main/java/leetcode/reversestring_344/Solution.java
package leetcode.reversestring_344;

/**
 * https://leetcode.com/problems/reverse-string/
 */
public class Solution {

    public String reverseString(char[] s) {
        int length = s.length;
        int left = 0;
        int right = length - 1;

        char[] reversed = new char[length];
        while (left <= right) {
            if (left == right) {
                reversed[left] = s[right];
                break;
            }

            reversed[left] = s[right];
            reversed[right] = s[left];

            left++;
            right--;
        }

        return String.valueOf(reversed);
    }

    // swap
    public String solution1(char[] s) {
        int start = 0;
        int end = s.length - 1;

        while (start < end) {
            char temp = s[start];
            s[start] = s[end];
            s[end] = temp;

            start++;
            end--;
        }

        return String.valueOf(s);
    }
}

File: src/main/java/leetcode/reversestring_344/Solution.kt
package leetcode.reversestring_344

fun reverseString(s: CharArray): Unit {
    var start = 0
    var end = s.size - 1

    while (start < end) {
        // also를 사용하면 모든 연산이 진행되기 전에는 값이 변하지 않는다.
        s[start] = s[end].also { s[end] = s[start] }

        start++
        end--
    }
}
